File: FlickrApp/ViewModels/SearchViewModel.m
//
//  SearchViewModel.m
//  FlickrApp
//
//  Created by aSqar on 26.11.2017.
//  Copyright © 2017 Askar Bakirov. All rights reserved.
//

#import "SearchViewModel.h"
#import "SearchAttemptViewModel.h"

@interface SearchViewModel()<RBQFetchedResultsControllerDelegate>

@property (nonatomic, strong) RACSubject *updatedContentSignal;
@property (nonatomic, strong) RBQFetchedResultsController *fetchedResultsController;

@end

@implementation SearchViewModel

#pragma mark - Public methods

-(instancetype)init
{
    self = [super init];
    if (self == nil)
        return nil;
    
    self.updatedContentSignal = [[RACSubject subject] setNameWithFormat:@"SearchResultViewModel updatedContentSignal"];
    
    @weakify(self)
    [self.didBecomeActiveSignal subscribeNext:^(id x) {
        @strongify(self);
        [self loadHistory];
    }];
    
    return self;
}

- (void) loadHistory
{
    [self.fetchedResultsController performFetch];
}

-(NSInteger)numberOfSections
{
    return [self.fetchedResultsController numberOfSections];
}

-(NSInteger)numberOfItemsInSection:(NSInteger)section
{
    return [self.fetchedResultsController numberOfRowsForSectionIndex:section];
}

-(SearchAttemptViewModel *) objectAtIndexPath:(NSIndexPath *)indexPath
{
    SearchAttempt *search = [self.fetchedResultsController objectAtIndexPath:indexPath];
    return [[SearchAttemptViewModel alloc] initWithSearchAttempt:search];
}

#pragma mark - Fetched results controller

- (RBQFetchRequest *) fetchRequest
{
    RLMSortDescriptor *sd1 = [RLMSortDescriptor sortDescriptorWithKeyPath:@"dateSearched" ascending:NO];
    NSArray *sortDescriptors = @[ sd1 ];
    
    RBQFetchRequest *fetchRequest = [RBQFetchRequest fetchRequestWithEntityName:@"SearchAttempt" inRealm:self.realm predicate:nil];
    [fetchRequest setSortDescriptors:sortDescriptors];
    return fetchRequest;
}

- (RBQFetchedResultsController *) fetchedResultsController
{
    if (_fetchedResultsController == nil) {
        _fetchedResultsController = [[RBQFetchedResultsController alloc] initWithFetchRequest:self.fetchRequest sectionNameKeyPath:nil cacheName:nil];
        [_fetchedResultsController setDelegate:self];
        [_fetchedResultsController performFetch];
    }
    
    return _fetchedResultsController;
}

- (void)controllerDidChangeContent:(RBQFetchedResultsController *)controller
{
    [(RACSubject *)self.updatedContentSignal sendNext:nil];
}

- (void)controllerWillChangeContent:(nonnull RBQFetchedResultsController *)controller
{
    
}

- (void)controller:(nonnull RBQFetchedResultsController *)controller
   didChangeObject:(nonnull RBQSafeRealmObject *)anObject
       atIndexPath:(nullable NSIndexPath *)indexPath
     forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(nullable NSIndexPath *)newIndexPath
{
    
}

- (void)controller:(nonnull RBQFetchedResultsController *)controller
  didChangeSection:(nonnull RBQFetchedResultsSectionInfo *)section
           atIndex:(NSUInteger)sectionIndex
     forChangeType:(NSFetchedResultsChangeType)type
{
    
}

@end

File: FlickrApp/ViewModels/SearchViewModel.swift
//
//  SearchViewModel.swift
//  FlickrApp
//
//  Created by aSqar on 26.11.2017.
//  Copyright © 2017 Askar Bakirov. All rights reserved.
//

import Realm
import RBQFetchedResultsController
import ReactiveCocoa

class SearchViewModel : BaseViewModel, RBQFetchedResultsControllerDelegate {

    // MARK: - Public methods

    private(set) var updatedContentSignal:RACSignal!
    
    private var _fetchedResultsController:RBQFetchedResultsController!
    var fetchedResultsController:RBQFetchedResultsController! {
        get { 
            if _fetchedResultsController == nil {
                _fetchedResultsController = RBQFetchedResultsController(fetchRequest:self.fetchRequest(), sectionNameKeyPath:nil, cacheName:nil)
                _fetchedResultsController.delegate = self
                _fetchedResultsController.performFetch()
            }

            return _fetchedResultsController
        }
        set { _fetchedResultsController = newValue }
    }

    override init() {
        super.init()
        
        self.updatedContentSignal = RACSubject()
        self.updatedContentSignal.name = "SearchResultViewModel updatedContentSignal"

        self.didBecomeActiveSignal.subscribeNext({ (x) in
            self.loadHistory()
            })

    }

    func loadHistory() {
        self.fetchedResultsController.performFetch()
    }

    func numberOfSections() -> Int {
        return self.fetchedResultsController.numberOfSections()
    }

    func numberOfItemsInSection(section:Int) -> Int {
        return self.fetchedResultsController.numberOfRows(forSectionIndex: section)
    }

    func objectAtIndexPath(indexPath:IndexPath!) -> SearchAttemptViewModel! {
        let search:SearchAttempt! = self.fetchedResultsController.object(at: indexPath) as! SearchAttempt
        return SearchAttemptViewModel(searchAttempt:search)
    }

    // MARK: - Fetched results controller

    func fetchRequest() -> RBQFetchRequest! {
        let sd1:RLMSortDescriptor! = RLMSortDescriptor(keyPath:"dateSearched", ascending:false)
        let sortDescriptors:[RLMSortDescriptor]! = [ sd1 ]

        let fetchRequest:RBQFetchRequest! = RBQFetchRequest(entityName:"SearchAttempt", in:self.realm(), predicate:nil)
        fetchRequest.sortDescriptors = sortDescriptors
        return fetchRequest
    }

    // `fetchedResultsController` has moved as a getter.

    func controllerDidChangeContent(_ controller:RBQFetchedResultsController) {
        (self.updatedContentSignal as! RACSubject).sendNext({ (x:Any!) in })
    }

    func controllerWillChangeContent(_ controller:RBQFetchedResultsController) {

    }

    func controller(_ controller:RBQFetchedResultsController, didChange anObject:RBQSafeRealmObject, at indexPath:IndexPath?, for type:NSFetchedResultsChangeType, newIndexPath:IndexPath?) {

    }

    func controller(_ controller:RBQFetchedResultsController, didChangeSection section:RBQFetchedResultsSectionInfo, at sectionIndex:UInt, for type:NSFetchedResultsChangeType) {

    }
}
